### PR TITLE
Fixing manager restore nemesis region issue

### PR DESCRIPTION
**Pull request by ShlomiBalalis**
### Testing

- [ ]

### PR pre-checks (self review)

- [x] I added the relevant `backport` labels
- [x] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevent to this change (if needed)

### CODE REVIEW

**Comment by ShlomiBalalis** [comment]
In run # 11 the nemesis passed several times: https://argus.scylladb.com/test/fed0b2b9-66f8-4cd7-bef0-c2603e5b4a75/runs?additionalRuns[]=b38e7583-97cb-4aac-b34e-3b337512478f
(it only ultimately failed because of spot termiation)

File: sdcm/tester.py [hunk 1]
```diff
@@ -958,11 +959,7 @@ def setUp(self):  # pylint: disable=too-many-branches,too-many-statements
             self.run_commit_log_check_thread(self.get_duration(None))
 
     def set_system_auth_rf(self, db_cluster=None):
-
-        if not db_cluster:
-            db_cluster = self.db_cluster
-        # No need to change system tables when running via scylla-cloud
-        # Also, when running a Alternator via scylla-cloud, we don't have CQL access to the cluster
```
Comment (vponomaryov) [inline sdcm/tester.py:965]: Comment removal looks redundant.

File: sdcm/tester.py [hunk 2]
```diff
@@ -958,11 +959,7 @@ def setUp(self):  # pylint: disable=too-many-branches,too-many-statements
...
+        db_cluster = db_cluster or self.db_cluster
```
Comment (vponomaryov) [inline sdcm/tester.py:962]: It looks redundant, because the `db_cluster` discovery gets handled by the `set_ks_rf` method.

**Review by vponomaryov — CHANGES_REQUESTED** [review]
Please, squash 2nd and 3rd commits into a single one.

File: sdcm/tester.py (hunk 2 above)
Comment (ShlomiBalalis) [reply sdcm/tester.py:962]: Yes, but you need it for the `if len(db_cluster.nodes) == 1` check

Comment (vponomaryov) [reply sdcm/tester.py:962]: Ok.

**Comment by ShlomiBalalis** [comment]
> Please, squash 2nd and 3rd commits into a single one. Also, need to fix unit tests.

> Comment removal looks redundant.

Done and done

**Comment by vponomaryov** [comment]
> > Please, squash 2nd and 3rd commits into a single one. Also, need to fix unit tests.
> 
> > Comment removal looks redundant.
> 
> Done and done

I don't see changes to unit tests.

Also, would be good to mention the https://github.com/scylladb/scylla-cluster-tests/issues/7122 in the main commit.

**Comment by ShlomiBalalis** [comment]
> > > Please, squash 2nd and 3rd commits into a single one. Also, need to fix unit tests.
> > 
> > 
> > > Comment removal looks redundant.
> > 
> > 
> > Done and done
> 
> I don't see changes to unit tests.

Yeah, test_temporary_replication_strategy_setter_rolls_back_on_exit needs fixing. On it.

> Also, would be good to mention, in the main commit, the bug that gets fixed here (#7122).

I'm not sure it will completely solve it, so I didn't think it should be mentioned atm

**Comment by vponomaryov** [comment]
> > Also, would be good to mention, in the main commit, the bug that gets fixed here (#7122).
> 
> I'm not sure it will completely solve it, so I didn't think it should be mentioned ATM

According to the description of the fix, it cover the root cause of the described bug.
So, I expect it to have 100% fix coverage.
Or am I missing something in this picture?

**Comment by ShlomiBalalis** [comment]
> > > Also, would be good to mention, in the main commit, the bug that gets fixed here (#7122).
> > 
> > 
> > I'm not sure it will completely solve it, so I didn't think it should be mentioned ATM
> 
> According to the description of the fix, it covers the root cause of the described bug. So, I expect it to have 100% fix coverage. Or am I missing something in this picture?

I'm not sure it's the only root cause. If it was I don't think only `3 out 4 10gb ones` was a thing. That's why I planned on fixing the region issue first, and then see if it completely solves the issue.

File: sdcm/tester.py [hunk 3]
```diff
@@ -974,21 +973,24 @@ def set_system_auth_rf(self, db_cluster=None):
         if self.test_config.REUSE_CLUSTER:
             self.log.debug("Skipping change rf of system_auth for reusing cluster")
             return
-        node = self.db_cluster.nodes[0]
+        self.set_ks_rf(keyspace="system_auth", db_cluster=db_cluster)
+
+    def set_ks_rf(self, keyspace, db_cluster=None):
```
Comment (fruch) [inline sdcm/tester.py:978]: Name is a bit vague.

It doesn't state to which replication factor it's gonna change and it's not in the parameters.

Also seems like this function belongs to Cluster and not Tester

File: sdcm/tester.py [hunk 4]
```diff
@@ -974,21 +973,24 @@ def set_system_auth_rf(self, db_cluster=None):
...
-        self.log.debug("system_auth description: %s", res.stdout)
-        self.log.info('repair system_auth keyspace ...')
+        NetworkTopologyReplicationStrategy(**datacenters).apply(node, keyspace)
+        res = node.run_cqlsh(f'DESC KEYSPACE "{keyspace}"', num_retry_on_failure=3)
+        self.log.debug("%s description: %s", keyspace, res.stdout)
+        self.log.info('repair %s keyspace ...', keyspace)
         for node in self.db_cluster.nodes:
-            node.run_nodetool(sub_cmd="repair -pr system_auth", timeout=MINUTE_IN_SEC * 20)
-        self.log.info('repair system_auth keyspace done')
+            node.run_nodetool(sub_cmd=f"repair -pr {keyspace}", timeout=MINUTE_IN_SEC * 20)
```
Comment (fruch) [inline sdcm/tester.py:992]: I'm not sure  repair is the side effect you want in restore nemesis, isn't that part of the restore itself ?

**Comment by vponomaryov** [comment]
Please, combine commit with unit tests updates and the logic change that caused the need to update it.

> > > > Also, would be good to mention, in the main commit, the bug that gets fixed here (#7122).
> > > 
> > > 
> > > I'm not sure it will completely solve it, so I didn't think it should be mentioned ATM
> > 
> > 
> > According to the description of the fix, it covers the root cause of the described bug. So, I expect it to have 100% fix coverage. Or am I missing something in this picture?
> 
> I'm not sure it's the only root cause. If it was I don't think only `3 out 4 10gb ones` was a thing. That's why I planned on fixing the region issue first, and then see if it completely solves the issue.

You may use `Ref: #foo`

File: sdcm/utils/replication_strategy_utils.py [hunk 5]
```diff
@@ -25,7 +25,7 @@ def get(cls, node: BaseNode, keyspace: str):
         return ReplicationStrategy.from_string(create_ks_statement)
 
     def apply(self, node: BaseNode, keyspace: str):
-        cql = f"ALTER KEYSPACE {keyspace} WITH replication = {self}"
+        cql = f'ALTER KEYSPACE "{keyspace}" WITH replication = {self}'
```
Comment (fruch) [inline sdcm/utils/replication_strategy_utils.py:28]: I suggest running it via cql_qoute function, which is intended exactly for those cases

Comment (fruch) [reply sdcm/utils/replication_strategy_utils.py:28]: Then you won't need to change all the unit tests

File: sdcm/tester.py (hunk 3 above)
Comment (ShlomiBalalis) [reply sdcm/tester.py:978]: > Name is a bit vague.

> It doesn't state to which replication factor it's gonna change and it's not in the parameters.

I don't disagree, I was just compiling with the previously written method

File: sdcm/tester.py (hunk 4 above)
Comment (ShlomiBalalis) [reply sdcm/tester.py:992]: It happens automatically after a restore, but you do need to repair a keyspace after altering it, do you not?
Also, I don't see why that would be an issue. It's not like we're repairing after restoring the data, only the schema.

Comment (fruch) [reply sdcm/tester.py:992]: for system_auth we are doing tha to make sure the auth data is on all nodes.

in this case isn't restore procedure puts date into each node ? i.e. if doing repair is part of the instructions for restore ? 
what the end user is going to do, and again where is the documentation for this flow, of data restored into different DC layout

**Comment by fruch** [comment]
@ShlomiBalalis 2 last standing issues:
- https://github.com/scylladb/scylla-cluster-tests/pull/7177#issuecomment-1921894603
- https://github.com/scylladb/scylla-cluster-tests/pull/7177#discussion_r1477312997

File: sdcm/tester.py (hunk 4 above)
Comment (ShlomiBalalis) [reply sdcm/tester.py]: Added an option to not repair after the alteration, and used it in my function call.
Also opened an issue for the docs to be amended: https://github.com/scylladb/scylla-manager/issues/3719

**Review by vponomaryov — APPROVED** [review]
LGTM

**Review by fruch — APPROVED** [review]
LGTM

**Pull request merged**